In [2]:
text_cols=['col49', 'col50', 'col51', 'col52', 'col53', 'col54', 'col55', 'col56', 'col204', 'col221', 'col222', 'col223', 'col224', 'col225', 'col226', 'col227', 'col228', 'col237', 'col238', 'col239', 'col240', 'col241', 'col242', 'col243', 'col244', 'col457', 'col458', 'col459', 'col460', 'col461', 'col462', 'col463', 'col464', 'col545', 'col546', 'col547', 'col548', 'col549', 'col550', 'col551', 'col552', 'col553', 'col554', 'col555', 'col556', 'col557', 'col558', 'col559', 'col560', 'col1046', 'col1047', 'col1048', 'col1049', 'col1050', 'col1051', 'col1052']
cat_cols=['год', 'сезон', 'месяц', 'client_id', 'col57', 'col58', 'col59', 'col60', 'col61', 'col62', 'col63', 'col64', 'col85', 'col86', 'col87', 'col88', 'col89', 'col90', 'col91', 'col92', 'col93', 'col94', 'col95', 'col96', 'col97', 'col98', 'col99', 'col100', 'col129', 'col130', 'col131', 'col132', 'col133', 'col134', 'col135', 'col136', 'col137', 'col138', 'col139', 'col140', 'col141', 'col142', 'col143', 'col144', 'col145', 'col146', 'col147', 'col148', 'col149', 'col150', 'col151', 'col152', 'col153', 'col154', 'col155', 'col156', 'col157', 'col158', 'col159', 'col160', 'col161', 'col162', 'col163', 'col164', 'col165', 'col166', 'col167', 'col168', 'col169', 'col170', 'col171', 'col172', 'col173', 'col174', 'col175', 'col176', 'col181', 'col182', 'col183', 'col184', 'col185', 'col186', 'col187', 'col188', 'col189', 'col190', 'col191', 'col192', 'col193', 'col194', 'col195', 'col196', 'col201', 'col202', 'col203', 'col205', 'col206', 'col207', 'col208', 'col209', 'col210', 'col211', 'col212', 'col213', 'col214', 'col215', 'col216', 'col465', 'col466', 'col467', 'col468', 'col469', 'col470', 'col471', 'col472', 'col513', 'col514', 'col515', 'col516', 'col517', 'col518', 'col519', 'col520', 'col521', 'col522', 'col523', 'col524', 'col525', 'col526', 'col527', 'col528', 'col529', 'col530', 'col531', 'col532', 'col533', 'col534', 'col535', 'col536', 'col537', 'col538', 'col539', 'col540', 'col541', 'col542', 'col543', 'col544', 'col561', 'col562', 'col563', 'col564', 'col565', 'col566', 'col567', 'col568', 'col569', 'col570', 'col571', 'col572', 'col573', 'col574', 'col575', 'col576', 'col577', 'col578', 'col579', 'col580', 'col581', 'col582', 'col583', 'col584', 'col585', 'col586', 'col587', 'col588', 'col589', 'col590', 'col591', 'col592', 'col593', 'col594', 'col595', 'col596', 'col597', 'col598', 'col599', 'col600', 'col601', 'col602', 'col603', 'col604', 'col605', 'col606', 'col607', 'col608', 'col609', 'col610', 'col611', 'col612', 'col613', 'col614', 'col615', 'col616', 'col793', 'col794', 'col795', 'col796', 'col797', 'col798', 'col799', 'col800', 'col1029', 'col1030', 'col1031', 'col1032', 'col1033', 'col1034', 'col1035', 'col1036', 'col1037', 'col1038', 'col1039', 'col1040', 'col1041', 'col1042', 'col1043', 'col1044', 'col1045', 'col1053', 'col1054', 'col1055', 'col1056', 'col1057', 'col1058', 'col1059', 'col1060', 'col1061', 'col1062', 'col1063', 'col1064', 'col1065', 'col1066', 'col1067', 'col1068', 'col1070', 'col1071', 'col1072', 'col1073', 'col1074', 'col1075', 'col1076', 'col1170', 'col1171', 'col1172', 'col1174', 'col1175', 'col1176', 'col1177', 'col1178', 'col1179', 'col1180', 'col1273', 'col1274', 'col1275', 'col1276', 'col1277', 'col1278', 'col1279', 'col1280', 'col1445', 'col1446', 'col1447', 'col1448', 'col1449', 'col1450', 'col1451', 'col1452', 'col1454', 'col1647', 'col1648', 'col1649', 'col1650', 'col1651', 'col1652', 'col1653', 'col1654', 'col2191', 'col2192', 'col2193', 'col2194', 'col2195', 'col2196', 'col2197', 'col2198']

In [3]:
cat_cols.remove('client_id')
cat_cols.remove('col1454')

In [4]:
from tqdm import tqdm

import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder, MinMaxScaler, StandardScaler, PowerTransformer
from sklearn.feature_extraction.text import TfidfVectorizer

from catboost import CatBoostClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
import lightgbm as lgb

from sklearn.model_selection import KFold, TimeSeriesSplit, GroupKFold, GroupShuffleSplit
from sklearn.metrics import roc_auc_score, precision_score, recall_score, f1_score, accuracy_score

import matplotlib.pyplot as plt
import seaborn as sns

In [5]:
df = pd.read_csv("../datasets/raw_data/dataset.csv", low_memory=False)

df['report_date'] = pd.to_datetime(df['report_date'])
df = df.sort_values('report_date')
df['год'] = df['report_date'].dt.year
df['сезон'] = (df['report_date'].dt.quarter % 4) + 1  # Добавьте 1, чтобы начать с 1
df['месяц'] = df['report_date'].dt.month
df.fillna(-1, inplace=True)
df.drop(columns=['report_date'], inplace=True)

for col in text_cols:
    df[col] = df[col].astype(str)
for col in cat_cols:
    df[col] = df[col].astype(str)
    le = LabelEncoder()
    df[col] = le.fit_transform(df[col])

In [30]:
X = df.drop(columns=['target','client_id','col1454'])
y = df['target']

# Convert categorical columns to 'category' type
for col in cat_cols:
    X[col] = X[col].astype('category')

# Transform text columns using TF-IDF
# Note: This is just one way to handle text features. Depending on your data, other methods might be more appropriate.
tfidf = TfidfVectorizer(max_features=15)  # Limiting to 100 features per text column for simplicity
for col in text_cols:
    tfidf_matrix = tfidf.fit_transform(X[col].fillna(''))
    tfidf_df = pd.DataFrame(tfidf_matrix.toarray(), columns=[f"{col}_{i}" for i in range(tfidf_matrix.shape[1])])
    X = pd.concat([X, tfidf_df], axis=1)
    X = X.drop(col, axis=1)

for col in ['col819', 'col820', 'col823', 'col824']:
    X[col] = X[col].replace('None', -1).astype(int)

In [64]:
avgrocs = {}
m = df['client_id']<3000
X_tr, y_tr = X[m], y[m]
X_vl, y_vl = X[~m], y[~m]
mms = MinMaxScaler((-2,2))
X_tr_ = mms.fit_transform(np.column_stack((y_tr, X_tr)))
X_vl_ = mms.transform(np.column_stack((np.full_like(y_vl, 0), X_vl)))
X_ = np.vstack((X_tr_, X_vl_))
gram = X_.T @ X_
diag_indices = np.diag_indices(gram.shape[0])
presicion = np.linalg.inv(gram + 1e5*np.eye(gram.shape[0]))
B = presicion / (-np.diag(presicion))
B[diag_indices] = 0
pred_tr, pred_vl = X_tr_@B[:,0], X_vl_@B[:,0]
roc_auc_score(y_vl, pred_vl)


0.7313444411411141

In [65]:
avgrocs = {}
m = df['client_id']<3000
X_tr, y_tr = X[m], y[m]
X_vl, y_vl = X[~m], y[~m]
model_cb = CatBoostClassifier(verbose=0, cat_features=cat_cols, iterations=1000, depth=3)
model_cb.fit(X_tr, y_tr)
print(f"{roc_auc_score(y_vl, model_cb.predict_proba(X_vl)[:,1]):.4f}")

0.7852


In [66]:
sX_vl = np.column_stack((model_cb.predict_proba(X_vl)[:,1], pred_vl))
sX_tr = np.column_stack((model_cb.predict_proba(X_tr)[:,1], pred_tr))

In [67]:
head = LogisticRegression().fit(sX_tr, y_tr)
print(f"{roc_auc_score(y_vl, head.predict_proba(sX_vl)[:,1]):.4f}")

0.7771


In [69]:

rocs_tr, rocs_vl = [], []
# for tr, vl in KFold(5, shuffle=True, random_state=0).split(X):
for tr, vl in KFold(5).split(X):
    X_tr, X_vl = X.iloc[tr], X.iloc[vl]
    y_tr, y_vl = y.iloc[tr], y.iloc[vl]

    model_cb = CatBoostClassifier(verbose=0, cat_features=cat_cols, depth=3)
    model_cb.fit(X_tr, y_tr)

    mms = MinMaxScaler((-2,2))
    X_tr_ = mms.fit_transform(np.column_stack((y_tr, X_tr)))
    X_vl_ = mms.transform(np.column_stack((np.full_like(y_vl, 0), X_vl)))
    X_ = np.vstack((X_tr_, X_vl_))
    gram = X_.T @ X_
    diag_indices = np.diag_indices(gram.shape[0])
    presicion = np.linalg.inv(gram + 1e4*np.eye(gram.shape[0]))
    B = presicion / (-np.diag(presicion))
    B[diag_indices] = 0
    pred_tr, pred_vl = X_tr_@B[:,0], X_vl_@B[:,0]

    sX_tr = np.column_stack((model_cb.predict_proba(X_tr)[:,1], pred_tr))
    sX_vl= np.column_stack((model_cb.predict_proba(X_vl)[:,1], pred_vl))
    head = LogisticRegression().fit(sX_tr, y_tr)

    rocs_tr.append(roc_auc_score(y_tr, head.predict_proba(sX_tr)[:,1]))
    rocs_vl.append(roc_auc_score(y_vl, head.predict_proba(sX_vl)[:,1]))
print(np.mean(rocs_tr), np.mean(rocs_vl))

0.9118231582464713 0.7983355809357818


| $\Lambda$ | record | month | client |
|---|---|---|---|
| 1e5 | 0.7461 | 0.7158 | __0.7402__ |
| 1e4 | 0.7892 | __0.7411__ | 0.7304 |
| 1e3 | 0.8076 | __0.7411__ | 0.7229 |
| 1e2 | __0.8096__ | 0.7340 | 0.7035 |


| model | record | month  | client |
|---|---|---|---|
| ease | 0.8096 | 0.7411 | 0.7402 |
| cb6 | 0.9053 | 0.8228 | 0.7675 |
| cb3 |  | 0.8134 | 0.7771 |
| stack3 | | 0.7983 |

In [68]:
imp = pd.DataFrame({'name':model_cb.feature_names_,'imp':model_cb.feature_importances_})
imp[imp['imp']>1e-2]

name       imp
112      col121  0.011244
119      col128  0.016753
131      col140  0.011157
139      col148  0.014696
214      col232  0.020685
...         ...       ...
6702  col556_24  0.015702
6706  col556_28  0.012599
6921  col558_43  0.047866
7055  col559_77  0.040611
7137  col560_59  0.013645

[463 rows x 2 columns]